File: api/telemetry/telemetry.go
// Package telemetry provides anonymous usage tracking for Runbooks.
// Telemetry is enabled by default but can be disabled via:
//   - Environment variable: RUNBOOKS_TELEMETRY_DISABLE=1
//   - CLI flag: --no-telemetry
//
// We collect minimal, anonymous data to improve Runbooks:
//   - Commands used (open, watch, serve)
//   - OS and architecture
//   - Runbooks version
//   - Error types (not content)
//
// We do NOT collect:
//   - Runbook content or file paths
//   - Variable values or script contents
//   - Personal identifiable information
//
// Learn more: https://runbooks.gruntwork.io/security/telemetry/
package telemetry

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"os/user"
	"runtime"
	"sync"
	"time"

	"github.com/mixpanel/mixpanel-go"
)

const (
	// EnvDisable is the environment variable to disable telemetry
	EnvDisable = "RUNBOOKS_TELEMETRY_DISABLE"

	// DocsURL is the documentation URL for telemetry
	DocsURL = "https://runbooks.gruntwork.io/security/telemetry/"
)

var (
	// MixpanelToken is the project token for Runbooks telemetry
	// This is set via ldflags at build time:
	//   go build -ldflags "-X runbooks/api/telemetry.MixpanelToken=your_token"
	// This is a public token - it only allows sending events, not reading data
	MixpanelToken = ""
)

var (
	// Global telemetry instance
	instance *Telemetry
	once     sync.Once
)

// Telemetry handles anonymous usage tracking
type Telemetry struct {
	enabled     bool
	version     string
	anonymousID string
	client      *mixpanel.ApiClient
	mu          sync.Mutex
}

// Config holds telemetry configuration for API responses
type Config struct {
	Enabled     bool   `json:"enabled"`
	AnonymousID string `json:"anonymousId"`
	Version     string `json:"version"`
}

// Init initializes the global telemetry instance
// Should be called once at startup with the application version
func Init(version string, disabledByFlag bool) {
	once.Do(func() {
		instance = &Telemetry{
			version: version,
		}

		// Check if telemetry is disabled via environment variable or flag
		if os.Getenv(EnvDisable) == "1" || os.Getenv(EnvDisable) == "true" || disabledByFlag {
			instance.enabled = false
			return
		}

		// Check if Mixpanel token is configured (set at build time via ldflags)
		if MixpanelToken == "" {
			instance.enabled = false
			return
		}

		instance.enabled = true
		instance.anonymousID = generateAnonymousID()

		// Initialize Mixpanel client
		instance.client = mixpanel.NewApiClient(MixpanelToken)
	})
}

// IsEnabled returns whether telemetry is enabled
func IsEnabled() bool {
	if instance == nil {
		return false
	}
	return instance.enabled
}

// GetConfig returns the telemetry configuration for API responses
func GetConfig() Config {
	if instance == nil {
		return Config{Enabled: false}
	}
	return Config{
		Enabled:     instance.enabled,
		AnonymousID: instance.anonymousID,
		Version:     instance.version,
	}
}

// Track sends an event to Mixpanel asynchronously
// This function never blocks and silently ignores errors
func Track(event string, properties map[string]any) {
	if instance == nil || !instance.enabled || instance.client == nil {
		return
	}

	// Run tracking in a goroutine to avoid blocking
	go func() {
		instance.mu.Lock()
		defer instance.mu.Unlock()

		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()

		// Add standard properties
		props := make(map[string]any)
		for k, v := range properties {
			props[k] = v
		}
		props["$os"] = runtime.GOOS
		props["$os_version"] = runtime.GOARCH
		props["version"] = instance.version
		props["time"] = time.Now().Unix()

		// Create and send the event
		eventData := instance.client.NewEvent(event, instance.anonymousID, props)

		// Silently ignore errors - telemetry should never impact user experience
		_ = instance.client.Track(ctx, []*mixpanel.Event{eventData})
	}()
}

// TrackCommand is a convenience function for tracking CLI command usage
func TrackCommand(command string) {
	Track("command_invoked", map[string]any{
		"command": command,
	})
}

// TrackError is a convenience function for tracking error types
// Note: We only track the error type, never the error message or content
func TrackError(errorType string) {
	Track("error_occurred", map[string]any{
		"error_type": errorType,
	})
}

// PrintNotice prints a telemetry notice to stdout
// This should be called on every command invocation when telemetry is enabled
func PrintNotice() {
	if !IsEnabled() {
		return
	}

	fmt.Printf("\n📊 Telemetry is enabled. Set %s=1 to opt out.\n", EnvDisable)
	fmt.Printf("   Learn more: %s\n\n", DocsURL)
}

// generateAnonymousID creates a stable, anonymous identifier for this machine
// The ID is a SHA-256 hash of hostname + username, making it:
// - Stable: Same ID across sessions on the same machine
// - Anonymous: Cannot be reversed to identify the user
// - Unique: Different for each machine/user combination
func generateAnonymousID() string {
	hostname, _ := os.Hostname()
	username := ""
	if u, err := user.Current(); err == nil {
		username = u.Username
	}

	// Create a hash of hostname + username
	data := fmt.Sprintf("runbooks:%s:%s", hostname, username)
	hash := sha256.Sum256([]byte(data))

	// Return first 16 bytes as hex (32 characters) - enough for uniqueness
	return hex.EncodeToString(hash[:16])
}

// Shutdown gracefully shuts down telemetry
// This is a no-op currently but reserved for future use
func Shutdown() {
	// Currently a no-op, but could be used to flush pending events
}

File: src/telemetry.ts
/**
 * Mixpanel telemetry.
 *
 * Singleton pattern: call init() once at startup, then track()/trackCommand()/trackError().
 * Telemetry is anonymous (SHA256 of hostname+username) and can be disabled via
 * RUNBOOKS_TELEMETRY_DISABLE env var or CLI flag.
 *
 * Collected: commands used, OS, arch, version, error types.
 * NOT collected: file content, paths, personal info.
 *
 * Docs: https://runbooks.gruntwork.io/security/telemetry/
 */
import * as os from "os"
import * as crypto from "crypto"
import { Effect } from "effect"
import type { TelemetryShape } from "./services/Telemetry.ts"

// ---------------------------------------------------------------------------
// Build-time / env constant
// ---------------------------------------------------------------------------

const MIXPANEL_TOKEN_FALLBACK = ""
const TELEMETRY_DOCS_URL = "https://runbooks.gruntwork.io/security/telemetry/"
const DISABLE_ENV_VAR = "RUNBOOKS_TELEMETRY_DISABLE"

// ---------------------------------------------------------------------------
// Singleton state
// ---------------------------------------------------------------------------

let enabled = false
let appVersion = "unknown"
let anonymousId: string | undefined
let mixpanelClient: ReturnType<typeof import("mixpanel")["init"]> | undefined

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Initialize telemetry. Must be called once before any tracking.
 *
 * Disabled when:
 * - `disabledByFlag` is true (CLI --no-telemetry)
 * - RUNBOOKS_TELEMETRY_DISABLE env var is set to any truthy value
 */
export function init(version: string, disabledByFlag: boolean): void {
  appVersion = version

  const envDisabled = isTruthyEnv(process.env[DISABLE_ENV_VAR])
  if (disabledByFlag || envDisabled) {
    enabled = false
    return
  }

  const token = process.env.MIXPANEL_TOKEN ?? MIXPANEL_TOKEN_FALLBACK
  if (!token) {
    enabled = false
    return
  }

  try {
    // Dynamic import not needed -- mixpanel is a regular CJS dep
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const Mixpanel = require("mixpanel") as typeof import("mixpanel")
    mixpanelClient = Mixpanel.init(token, { protocol: "https" })
    anonymousId = generateAnonymousId()
    enabled = true
  } catch {
    // If mixpanel fails to load (e.g. missing dep in dev), silently disable.
    enabled = false
  }
}

/** Whether telemetry is currently active. */
export function isEnabled(): boolean {
  return enabled
}

/** Returns config info suitable for an API health-check response. */
export function getConfig(): { enabled: boolean; token: string | undefined } {
  return {
    enabled,
    token: enabled ? (process.env.MIXPANEL_TOKEN ?? MIXPANEL_TOKEN_FALLBACK) : undefined,
  }
}

/**
 * Track an event with optional properties. Fire-and-forget: never throws,
 * never blocks the caller.
 */
export function track(event: string, properties?: Record<string, unknown>): void {
  if (!enabled || !mixpanelClient) return

  try {
    mixpanelClient.track(event, {
      distinct_id: anonymousId,
      version: appVersion,
      os: process.platform,
      arch: process.arch,
      node_version: process.version,
      ...properties,
    })
  } catch {
    // Silently swallow -- telemetry must never disrupt the app.
  }
}

/** Convenience: track a CLI command invocation. */
export function trackCommand(command: string): void {
  track("command", { command })
}

/** Track error type (NOT content) for aggregate error reporting. */
export function trackError(errorType: string): void {
  track("error", { error_type: errorType })
}

/**
 * Generate a stable anonymous ID from hostname + username.
 * SHA-256 ensures no PII is transmitted.
 */
export function generateAnonymousId(): string {
  const hostname = os.hostname()
  const username = os.userInfo().username
  return crypto.createHash("sha256").update(`${hostname}:${username}`).digest("hex")
}

/** Print opt-out notice to stderr (shown once on first run). */
export function printNotice(): void {
  console.error(
    [
      "",
      "Gruntwork Runbooks collects anonymous usage telemetry to improve the product.",
      `To learn more or opt out, visit: ${TELEMETRY_DOCS_URL}`,
      `To disable: set ${DISABLE_ENV_VAR}=1`,
      "",
    ].join("\n"),
  )
}

/**
 * Gracefully flush pending events. Call on process exit.
 * Returns a promise that resolves after a short timeout so we don't
 * block shutdown indefinitely.
 */
export function shutdown(): Promise<void> {
  if (!enabled || !mixpanelClient) return Promise.resolve()

  // Mixpanel Node SDK doesn't expose a flush/close, so we just give a brief
  // window for in-flight HTTP requests to complete.
  return new Promise((resolve) => setTimeout(resolve, 500))
}

// ---------------------------------------------------------------------------
// Effect-based service implementation
// ---------------------------------------------------------------------------

/**
 * Creates a TelemetryShape backed by the singleton above.
 * Useful for providing the Telemetry service in the Effect context.
 */
export function makeTelemetryService(): TelemetryShape {
  return {
    track: (event, properties) =>
      Effect.sync(() => track(event, properties)),
    trackCommand: (command) =>
      Effect.sync(() => trackCommand(command)),
    trackError: (errorType) =>
      Effect.sync(() => trackError(errorType)),
    isEnabled: () =>
      Effect.sync(() => isEnabled()),
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isTruthyEnv(value: string | undefined): boolean {
  if (!value) return false
  const lower = value.toLowerCase()
  return lower === "1" || lower === "true" || lower === "yes"
}
